### DISCUSSION TITLE
[ ] works in wrap(entity).assign() but never produce SQL update queries

### DISCUSSION THREAD

**Discussion by sonicgamer0721** [mikro-orm/mikro-orm#3496] [Q&A]
**Describe the bug**
Hi, I wrote this code to update User entity from MariaDB by RESTful API, and I found something wired:
When I wanted to remove all roles, I used **roles=[]** in this code, but em.flush() did not produce update SQL queries to the database.
I need help, please.

**Stack trace**
```typescript
// User entity
@Entity({ tableName: 'core_users' })
export class User {
  @PrimaryKey()
  id: number;

  @Unique()
  @Property()
  account: string;

  @Property()
  name: string;

  @ManyToMany(() => Role, (role) => role.users)
  roles = new Collection<Role>(this);
}

// The Code
const em = this.em;

for await (const updateUserDto of updateUserDtos) {
  const { id, roles, ...dto } = updateUserDto;
  const user = await em.findOne(User, { id });
  wrap(user).assign({
    ...dto,
    roles: roles.map((id) => em.getReference(Role, id)),
  });
}

await em.flush();
```

**To Reproduce**
1. Download the source code (This is SQLite version) and extract it: [https://drive.google.com/file/d/1_atk7ZSFwOTnVURnVRgEpqHbzQlGmquJ/view?usp=sharing](url)
2. Run **npm install** and **npm run start**
3. Open http://127.0.0.1:3000/api in browser, you will see Swagger page.
4. Execute patch to /api/core/user with Request body: `[ { "id": 1, "roles": [] } ]` , from console you can see that no update queries were produced .
5. Execute patch to /api/core/user with Request body: `[ { "id": 1, "roles": [1, 3] } ]` and there will be update queries being produced.

**Expected behavior**
I hope **roles=[]** works fine when I want to remove all roles.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.15.0 |
| nestJS | 9.0.11 |
| typescript | 4.8.2 |
| mikro-orm | 5.3.1 |
| @mikro-orm/sqlite | 5.3.1 |

**Comment by B4nan** [comment]
Thats quite complex repro, I would expect a single script for this, not a nestjs application. Can you try to isolate it a bit? Also instead of uploading to google drive, please publish it on GH (or just past it as a text if its a single file).

edit: I guess it will be fine this way, but next time I would appreciate much simpler reproductions, e.g. I dont want to traverse nested folders and read 3 files to see what can be done in a single script within 50 lines of code :]

**Comment by sonicgamer0721** [comment]
OK. I got it. Thanks for your advice.

**Comment by B4nan** [comment]
You are trying to reset inverse side of the M:N relation, its probably caused by that. This can work only if all the items in that collection were populated. And seeing you do just `const user = await em.findOne(User, { id })`, they are clearly not populated.

Would be good to have a validation for this at least. Try this instead:

```ts
const user = await em.findOne(User, { id }, { populate: ['roles] });
wrap(user).assign({
	...dto,
	roles: roles.map((id) => em.getReference(Role, id)),
});
```

**Comment by sonicgamer0721** [comment]
Hi, B4nan. Thanks for your helping. I tried your advice, add `{ populate: ['roles'] }` and still no update queries produced.
Then I tried to use this code:
```typescript
const user = await em.findOne(User, { id }, { populate: ['roles'] });
user.roles.removeAll();
```
Same problem!
If I try to isolate my code and upload to Github, does that help?
